ADRs/2022-12-08-PropertyWrappers-ADR.md

```diff
@@ -0,0 +1,184 @@
+            ...
+        }
+    }
+}
+```
+
+The solution utilizes the same technique that powers SwiftUI's `@Published` property behavior, where the property wrapper parent can be accessed in a static subscript to add observation behavior to the wrapped value.
+
+Benefits
+
+- Removes 9 lines of boilerplate code (Solves problem #1)
+- Removes potential for incorrect state subscription (Solves problem #2)
+- Removes potential for memory leaks in state observation (Solves problem #3)
+- Invokes `render()` after the state value changes, ensuring `state` and `container.state` are both safe to access in the rendering code. (Solves problem #4)
+- Allows custom state subscription to be possible if required (by opting out of the `@AutoRendered` property)
+
+Drawbacks
+
+- Forgetting to apply `@AutoRendered` to the `container` property will result in runtime behavior issues (Property wrapper usage cannot be enforced by the compiler, as is the standard for all property wrappers like `@State`, `@ObservedObject`, etc.)
+
+## Alternatives Considered
+
+Two other solutions were considered which provide similar benefits while making the VSM framework more ergonomic.
+
+### ViewState Property Wrapper
+
+This solution focuses on addressing the problems in question while improving the ergonomics of the SwiftUI VSM view code. It does so by introducing a `@ViewState` property wrapper which encapsulates the `StateContainer` entirely from the engineer. It also provides the same convenient auto-observing behavior as `@AutoRender`, allowing it to be used for both SwiftUI and UIKit.
+
+All `StateContainer` properties and functions would then be accessed directly from the `@ViewState` wrapper by way of the `_` underscore prefix.
+
+Example Usage
+
+```swift
+struct SomeView: View {
+    @ViewState var state: SomeViewState
+    ...
+    var body: some View {
+        Button("\(state.isEnabled)") {
+            _state.observe(state.toggle(isEnabled: !state.isEnabled))
+        }
+        .onChange(of: state.isEnabled) { isEnabled in 
+            print(isEnabled)
+        }
+        .onReceive(_state.publisher) { newState in 
+            print(newState)
+        }
+    }
+}
+
+// Enables simple default initializer:
+SomeView(state: .init(SomeViewState()))
+```
+
+As you can see from the example, the `ViewStateRendering` protocol is no longer necessary.
+
+Benefits
+
+- Solves all the problems that `@AutoRendered` solves
+- Works for both UIKit and SwiftUI
+- More ergonomic for declaring the view state property on a view
+- More ergonomic for instantiating a view
+- No longer requires a protocol
+- The developer no longer has to choose between `@ObservedObject` and `@StateObject` for the `container` property in SwiftUI
+
+Drawbacks
+
+- The `_` prefix is generally unpopular for such a use case, causing much nose-wrinkling
+- Extra typing (`_state...`) is required on some lines because `observe` and `bind` are no longer accessible directly on the view
```

Comment:
I think this might actually be an improvement by making what is being observed more explicit.  Even in the existing VSM implementation, I find it more natural to use `container.observe` than simply `observe` for this reason.